File: libraries/fary/ary_loc.f
      SUBROUTINE ARY_LOC( IARY, LOC, STATUS )
*+
*  Name:
*     ARY_LOC

*  Purpose:
*     Obtain an HDS locator for an array.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL ARY_LOC( IARY, LOC, STATUS )

*  Description:
*     The routine returns an HDS locator for the data object referred to
*     by the supplied ARY identifier.

*  Arguments:
*     IARY = INTEGER (Given)
*        Array identifier.
*     LOC = CHARACTER * ( DAT__SZLOC ) (Returned)
*        The HDS locator. It should be annulled using DAT_ANNUL when no
*        longer needed. A value of DAT__NOLOC will be returned if an
*        error occurs.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Copyright:
*     Copyright (C) 2010 Science & Technology Facilities Council.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 59 Temple Place,Suite 330, Boston, MA
*     02111-1307, USA

*  Authors:
*     DSB: David S Berry (JAC)
*     {enter_new_authors_here}

*  History:
*     20-OCT-2010 (DSB):
*        Original version.
*     {enter_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'ARY_PAR'          ! ARY_ public constants
      INCLUDE 'ARY_CONST'        ! ARY_ private constants

*  Global Variables:
      INCLUDE 'ARY_DCB'          ! ARY_ Data Control Block
*        DCB_LOC( ACB_MXDCB ) = CHARACTER * ( DAT__SZLOC ) (Read)
*           Locator for the data object.

      INCLUDE 'ARY_ACB'          ! ARY_ Access Control Block
*        ACB_IDCB( ARY__MXACB ) = INTEGER (Read)
*           Index to data object entry in the DCB.

*  Arguments Given:
      INTEGER IARY

*  Arguments Returned:
      CHARACTER * ( DAT__SZLOC ) LOC

*  Status:
      INTEGER STATUS             ! Global status

*  Local variables:
      INTEGER IACB               ! Index to ACB entry
      INTEGER IDCB               ! Index to data object entry in the DCB

*.

*  Initialise
      LOC = DAT__NOLOC

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Import the array identifier.
      CALL ARY1_IMPID( IARY, IACB, STATUS )
      IF ( STATUS .EQ. SAI__OK ) THEN

*  Get the DCB index for the data object.
         IDCB = ACB_IDCB( IACB )

*  Clone the data object locator.
         CALL DAT_CLONE( DCB_LOC( IDCB ), LOC, STATUS )

      END IF

*  If an error occurred, then report context information and call the
*  error tracing routine.
      IF ( STATUS .NE. SAI__OK ) THEN
         CALL ERR_REP( 'ARY_LOC_ERR', 'ARY_LOC: Error obtaining an '//
     :                 'HDS locator for an array.', STATUS )
         CALL ARY1_TRACE( 'ARY_LOC', STATUS )
      END IF

      END

File: libraries/ary/aryLoc.c
#include "sae_par.h"
#include "ary1.h"
#include "star/hds.h"
#include "mers.h"

void aryLoc( Ary *ary, HDSLoc **loc, int *status ) {
/*
*+
*  Name:
*     aryLoc

*  Purpose:
*     Obtain an HDS locator for an array.

*  Synopsis:
*     void aryLoc( Ary *ary, HDSLoc **loc, int *status )

*  Description:
*     This function returns an HDS locator for the data object referred to
*     by the supplied ARY identifier.

*  Parameters:
*     ary
*        Array identifier.
*     loc
*        Returned holding the HDS locator. It should be annulled using
*        datAnnul when no longer needed. A value of NULL will be returned
*        if an error occurs.
*     status
*        The global status.

*  Copyright:
*      Copyright (C) 2017 East Asian Observatory
*      All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     03-JUL-2017 (DSB):
*        Original version, based on equivalent Fortran routine by RFWS.

*-
*/

/* Local variables: */
   AryACB *acb;
   AryDCB *dcb;

/* Initialise */
   *loc = NULL;

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Import the array identifier. */
   acb = (AryACB *) ary1Impid( ary, 1, 1, 1, status );
   if( *status == SAI__OK ){

/* Get the DCB index for the data object. */
      dcb = acb->dcb;

/* Clone the data object locator. */
      datClone( dcb->loc, loc, status );
   }

/* If an error occurred, then report context information and call the error
   tracing routine. */
   if( *status != SAI__OK ){
      errRep( " ", "aryLoc: Error obtaining an HDS locator for an array.",
              status );
      ary1Trace( "aryLoc", status );
   }

}
